File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        left = 0
        charset = set()
        max_length = 0

        for right in range(n):
            while s[right] in charset:
                charset.remove(s[left])
                left += 1
            charset.add(s[right])
            max_length = max(max_length, right - left + 1)
        
        return max_length

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> charSet;
        int left = 0;
        int max_length = 0;
        
        for (int right = 0; right < s.size(); ++right){
            while (charSet.find(s[right]) != charSet.end()){
                charSet.erase(s[left]);
                left++;
            }

            charSet.insert(s[right]);
            max_length = max(max_length, right - left + 1);
        }

        return max_length;
    }
};